### PR TITLE
Add hyperparameter config, use generic training interfaces

### PR DESCRIPTION
**Pull request by mcgibbon**
Currently, sklearn and keras training have separate pre-processing logic before their fit routines are called. Their configuration also loosely types the keyword arguments, by accepting kwargs. This opens the possibility for typos in configuration to go unnoticed, and makes it hard to document what options are available. It also puts model-specific concerns into the top-level routines, instead of being able to dispatch those concerns to Estimator classes.

This PR adds the concept of a hyperparameter configuration class for an Estimator, which contains all of the init arguments for that estimator. This is integrated into the TrainingConfig, and is the only part of the configuration that differs between different estimators.

Added public API:
- Added OptimizerConfig and RegularizerConfig which replace the current arguments for optimizer and regularizer on DenseModel

Refactored public API:
- Removed DenseTrainingConfig and SklearnTrainingConfig, added DenseHyperparameters and RandomForestHyperparameters which strongly type the hyperparameter configuration
- Added hyperparameter class to the API for `register_estimator`
- Renamed `get_config_class` to `get_hyperparameter_class`
- Predictor now takes `**kwargs` in its API to satisfy mypy, but raises a TypeError if any are passed

Significant internal changes:
- Modified Estimator classes to take a single "hyperparameters" dataclass argument, instead of multiple keyword arguments
- added RandomForest estimator which wraps the existing SklearnWrapper estimator, and also provides functionality from now-deleted `sklearn/_train.py`
- Removed target_scaler and n_jobs from the SklearnWrapper init, as there was no way to set these from configuration

- [ ] Tests added

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/fv3fit/train.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/train.py:54]: Note the central purpose of this PR is to be able to refactor this function like this.

File: external/fv3fit/fv3fit/sklearn/_wrapper.py
Comment (nbren12) [inline external/fv3fit/fv3fit/sklearn/_wrapper.py:42]: Nice. I like this way of providing a clean interface to the overly complex sklearn code. It might be worth renaming this module to `_random_forest.py` and making `RegressorEnsemble` and `SklearnWrapper` private APIs.

**Review by nbren12 — COMMENTED** [review]
Looking good! I'll let someone else review more closely. Some general comments/requests:

-  can you shorten the PR title if you plan to make it a commit message?
-  There are a lot of `Optional` configurations, and each one of these requires some conditional logic to disambiguate. You could use more concrete defaults in places to avoid this.

File: external/fv3fit/fv3fit/keras/_models/models.py
Comment (brianhenn) [inline external/fv3fit/fv3fit/keras/_models/models.py:360]: Leftover?

File: external/fv3fit/fv3fit/keras/_training.py
Comment (brianhenn) [inline external/fv3fit/fv3fit/keras/_training.py:23]: Haven't looked through the tests enough to see why this function is still needed, but it would be nice to be able to get rid of it since it's no longer used by the training routine. OTOH this PR is pretty big already...

File: external/fv3fit/fv3fit/_shared/config.py
Comment (brianhenn) [inline external/fv3fit/fv3fit/_shared/config.py:68]: Couldn't this be `dacite.from_dict(cls, **kwargs)`?

File: external/fv3fit/fv3fit/keras/_models/models.py
Comment (brianhenn) [inline external/fv3fit/fv3fit/keras/_models/models.py:321]: This is great to be able to dump this estimator's hyperparameters in a self-defining way.

**Review by brianhenn — COMMENTED** [review]
Thanks @mcgibbon this is going to be great. I found a number of small issues, and obviously need to make sure the e2e tests can pass since there is some coupling between fv3fit and the offline diags.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/_shared/config.py:208]: For flexibility in the future could you also add defaults `{"max_samples": Union[int, float]=None, "bootstrap": bool=True, "criterion": str="mse", "warm_start": bool=True}

The last especially is needed for removing `RegressorEnsemble`.

File: external/fv3fit/fv3fit/sklearn/_wrapper.py
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/sklearn/_wrapper.py:51]: I think in the current use case (using `RegressorEnsemble` as a hacky forest and 1 tree per sklearn RF) neither the higher-up `SklearnWrapper` n_jobs nor the sklearn model hyperparameter n_jobs are doing any parallelization of subsetted data after splits. Since there's no difference between picking one of those to set right now I think it would be easier for the future PR that removes the `RegressorEnsemble` to have `n_jobs` set as the `RandomForestRegressor` hyperparameter in this PR, so that this code doesn't have to get adjusted later. At least I think that is how the parallelization is currently working (or not working).

Comment (brianhenn) [reply external/fv3fit/fv3fit/sklearn/_wrapper.py:51]: Isn't the current parallel backend context in `predict` parallelizing calls over the RFs in the `RegressorEnsemble` if `n_jobs = -1`?

File: external/fv3fit/fv3fit/sklearn/_wrapper.py
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/sklearn/_wrapper.py:245]: Oh, good catch. I think this will cause an error when trying to use `.predict` from a loaded model.

File: external/fv3fit/fv3fit/typing.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/typing.py:6]: Yes, there's no built-in type so you have to make a Protocol object manually for hinting. I added a link to SO.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (brianhenn) [reply external/fv3fit/fv3fit/_shared/config.py:221]: OK as discussed that's fine

File: external/fv3fit/fv3fit/sklearn/_wrapper.py
Comment (brianhenn) [reply external/fv3fit/fv3fit/sklearn/_wrapper.py:51]: OK as discussed it's not actually doing any parallelism currently so we can delete that context and follow @AnnaKwa's suggestion

**Comment by AnnaKwa** [comment]
Thanks Jeremy, the new interface is much cleaner! The responses to the couple of comments I left LGTM.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (brianhenn) [inline external/fv3fit/fv3fit/_shared/config.py:370]: update function type signature to reflect

**Review by brianhenn — COMMENTED** [review]
Thanks I think my issues have been addressed. Happy to approve once tests are passing.

File: external/fv3fit/fv3fit/sklearn/_wrapper.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/sklearn/_wrapper.py:41]: Also, this has exposed that offline_ml_diags/_plot_input_sensitivity.py is tightly coupled to fv3fit's internals, because it type hints specific model types rather than having a jacobian/feature importances method defined. I'll work on that in another PR.

**Comment by mcgibbon** [comment]
Integration tests passed https://app.circleci.com/pipelines/github/VulcanClimateModeling/fv3net/8551/workflows/c46866c3-1f89-467c-b87b-457ab89c69a1

**Review by brianhenn — APPROVED** [review]
Thanks Jeremy!

**Pull request merged**
